agent/exploits/common.py

```diff
@@ -1,29 +1,197 @@
+        metadata.append(
+            agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                metadata_type=agent_report_vulnerability_mixin.MetadataType.URL,
+                value=full_url,
+            )
+        )
+        asset = domain_asset.DomainName(name=prepare_domain_asset(target_url))
+
+    if target.port is not None or (ip is not None and port is not None):
+        metadata_type = agent_report_vulnerability_mixin.MetadataType.PORT
+        metadata_value = str(target.port) if target.port is not None else port
+        if metadata_value is not None:
+            metadata.append(
+                agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                    metadata_type=metadata_type, value=metadata_value
+                )
+            )
+
+    return agent_report_vulnerability_mixin.VulnerabilityLocation(
+        asset=asset, metadata=metadata
+    )
+
+
+def prepare_domain_asset(url: str) -> str:
+    """Prepares the domain asset object for the given URL.
+
+    Args:
+    url: The URL to extract the domain from.
+
+    Returns:
+    Optional[domain_asset.DomainName]: A domain asset
+    """
+    if url is None:
+        return ""
+
+    canonized_domain = tld.get_tld(
+        url, as_object=True, fix_protocol=True, fail_silently=True
+    )
+
+    if canonized_domain is None:
+        return parse.urlparse(url).netloc
+
+    tld_domain = cast(tld.Result, canonized_domain)
+    result_neloc = tld_domain.parsed_url.netloc
+    if ":" in result_neloc:
+        asset = result_neloc.split(":")[0]
+    else:
+        asset = result_neloc
+
+    return asset
+
+
+def compute_dna(
+    vulnerability_title: str,
+    vuln_location: agent_report_vulnerability_mixin.VulnerabilityLocation,
+) -> str:
+    """Compute a deterministic, debuggable DNA representation for a vulnerability.
+
+    Args:
+        vulnerability_title: The title of the vulnerability.
+        vuln_location: The location of the vulnerability.
+
+    Returns:
+        A deterministic JSON representation of the vulnerability DNA.
+    """
+    dna_data: dict[str, Any] = {"title": vulnerability_title}
+
+    if vuln_location is not None:
+        location_dict: dict[str, Any] = vuln_location.to_dict()
+        sorted_location_dict = sort_dict(location_dict)
+        dna_data["location"] = sorted_location_dict
+
+    return json.dumps(dna_data, sort_keys=True)
+
+
+def sort_dict(d: dict[str, Any] | list[Any]) -> dict[str, Any] | list[Any]:
+    """Recursively sort dictionary keys and lists within.
+
+    Args:
+        d: The dictionary or list to sort.
+
+    Returns:
+        A sorted dictionary or list.
+    """
+    if isinstance(d, dict):
+        return {k: sort_dict(v) for k, v in sorted(d.items())}
+    if isinstance(d, list):
+        return sorted(
+            d,
+            key=lambda x: json.dumps(x, sort_keys=True)
+            if isinstance(x, dict)
+            else str(x),
+        )
+    return d
```

Comment:
```suggestion
    return d
```
At runtime, d can be of any type, so explicit handling for cases where d is neither a dict nor a list is better than just returning it.

If d is not a list or a dict, the function would return a different type. If this is acceptable, make it explicit by returning Any.